### PR TITLE
UnorderedMap(space instance): proposal for #6067

### PR DESCRIPTION
**Pull request by romintomasetti**
This is a proposal for:
- #6067

1. I needed to modify both `UnorderedMap` and `Bitset` to ensure no global fencing occurs.
2. Also, the member views are now always named either as `<user provided object label> - <member hardcoded label>` or if no user label is given then a default is used. This is all done using the new `with_updated_label`.
3. Now, `Bitset` is not default initialized in the default constructor (size was before hardcoded to zero)
4. Miscellaneous changes in `Bitset` due to `clang format` 
5. 2 tests in `TestUnorderedMap` were not running using the device though I think they should so I modified them

- [ ] Check that `with_updated_label` is semantically OK and check where it should be (not sure it should be defined in `Kokkos_UnorderMap.hpp` nor in `Kokkos_Bitset.hpp`)

### CODE REVIEW

**Comment by Rombur** [comment]
OK to test

**Comment by romintomasetti** [comment]
@dalg24 Could you give your input on this one? Iguess at least thee aspects require your inspection:
1. `with_updated_label`
2. if `Bitset` should not be initialized by default (before, it was always initialized to a zero size bitset, now the default constructor leaves all members uninitialized)
3. I remove `iostream` from some files. It is ok except for `THREADS` backend. I wonder why `std` includes come after `Kokkos` includes in `Kokkos_core.cpp`. I guess the missing `iostream` should be added on the top of `Kokkos_core.cpp`

Thanks!

**Comment by romintomasetti** [comment]
@Rombur or @dalg24 Could you please have a look :smile: Thanks!

File: core/src/impl/Kokkos_Core.cpp [hunk 1]
```diff
@@ -29,7 +31,6 @@
 #include <algorithm>
 #include <cctype>
 #include <cstring>
-#include <iostream>
```
Comment (Rombur) [inline core/src/impl/Kokkos_Core.cpp:32]: Why did you change this?

File: core/src/impl/Kokkos_ExecPolicy.cpp [hunk 2]
```diff
@@ -18,9 +18,11 @@
 #define KOKKOS_IMPL_PUBLIC_INCLUDE
 #endif
 
-#include <Kokkos_Core.hpp>
+#include <iostream>
```
Comment (Rombur) [inline core/src/impl/Kokkos_ExecPolicy.cpp:21]: Why did you add `<iostream>` and move `<Kokkos_Core.hpp>`?

File: core/src/Threads/Kokkos_ThreadsExec.hpp [hunk 3]
```diff
@@ -20,8 +20,9 @@
 #include <Kokkos_Macros.hpp>
 
 #include <cstdio>
-
+#include <ostream>
```
Comment (Rombur) [inline core/src/Threads/Kokkos_ThreadsExec.hpp:23]: Why did you add `<ostream>`?

**Review by Rombur — COMMENTED** [review]
What's going on with all the `<iostream>` changes?

File: core/src/Threads/Kokkos_ThreadsExec.hpp (hunk 3 above)
Comment (romintomasetti) [reply core/src/Threads/Kokkos_ThreadsExec.hpp:23]: Because it is needed by `Kokkos::Impl::ThreadExec::print_configuration(std::ostream&, const bool detail = false);` defined below.

Comment (romintomasetti) [reply core/src/Threads/Kokkos_ThreadsExec.hpp:23]: Maybe I could just use `<iosfwd>` in this case :smile:

File: core/src/impl/Kokkos_Core.cpp (hunk 1 above)
Comment (romintomasetti) [reply core/src/impl/Kokkos_Core.cpp:32]: Because system headers come first, as per https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes. But I can revert this kind of changes if you prefer :wink:

**Comment by romintomasetti** [comment]
> What's going on with all the `<iostream>` changes?

So what I wanted to do is remove `#include <iostream>` from headers that a user would use, and include it only when necessary. Usually compiling with `<iostream>` and other big system includes slows down the compilation. So I wanted to remove it from Kokkos headers, and so I needed to move it to Kokkos `.cpp` files where needed.

File: containers/unit_tests/TestUnorderedMap.hpp [hunk 4]
```diff
@@ -441,17 +441,49 @@ TEST(TEST_CATEGORY, UnorderedMap_valid_empty) {
   ASSERT_TRUE(n.is_allocated());
 }
 
+/**
+ * This helper is needed because NVCC does not like extended lambdas
+ * in private member functions.
+ * Google Test bodies are private member functions. So it is incompatible.
+ * See also https://github.com/google/googletest/issues/4104.
+ */
```
Comment (romintomasetti) [inline containers/unit_tests/TestUnorderedMap.hpp:449]: Hi @dalg24 ! What do you think of this issue? Do you know the reason why Cuda requires that extended lambda cannot be defined in non-public members?

File: containers/unit_tests/TestUnorderedMap.hpp [hunk 5]
```diff
@@ -441,17 +441,49 @@ TEST(TEST_CATEGORY, UnorderedMap_valid_empty) {
...
+  void insert(Args &&... args) const {
+    static_assert(sizeof...(Args) > 1, "Prefer the single value version");
+    constexpr size_t size = sizeof...(Args);
+    Kokkos::Array<typename map_type::key_type, size> values{
+        std::forward<Args>(args)...};
+    Kokkos::parallel_for(
+        Kokkos::RangePolicy<typename map_type::execution_space>(0, size),
+        KOKKOS_CLASS_LAMBDA(const int i) { this->m_map.insert(values[i]); });
+  }
+};
```
Comment (romintomasetti) [inline containers/unit_tests/TestUnorderedMap.hpp:476]: Hi @dalg24 ! I couldn't use the same name for both functions and use something like `std::enable_if_t<(sizeof...(Args) > 1)>` to differentiate between the single and multiple values specializations, because Cuda does not like it. I think we have a similar situation in our code, and it works because we use `requires` instead. Do you know any tweak here or are you fine with one `insert` and one `insert_single`?

File: core/src/impl/Kokkos_Core.cpp (hunk 1 above)
Comment (romintomasetti) [reply core/src/impl/Kokkos_Core.cpp:32]: Ok, I'll revert that change :wink:

Comment (romintomasetti) [reply core/src/impl/Kokkos_Core.cpp]: Reverted.

File: containers/unit_tests/TestUnorderedMap.hpp (hunk 4 above)
Comment (Rombur) [reply containers/unit_tests/TestUnorderedMap.hpp:449]: It's just one of the restrictions of using lambda with CUDA https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#extended-lambda-restrictions

File: containers/unit_tests/TestBitset.hpp [hunk 6]
```diff
@@ -156,7 +156,7 @@ void test_bitset() {
   {
     unsigned ts = 100u;
     bitset_type b1;
-    ASSERT_TRUE(b1.is_allocated());
+    ASSERT_FALSE(b1.is_allocated());
```
Comment (Rombur) [inline containers/unit_tests/TestBitset.hpp:159]: So Bitset is always allocated now even if the size is zero?

Comment (romintomasetti) [reply containers/unit_tests/TestBitset.hpp:159]: I think it's the contrary. Before this PR, `Bitset` default construction would allocate its member with a zero-size (`Bitset(unsigned arg_size = 0u)`). Now, the default constructor does nothing, so by default it is not allocated. However, I think the behavior when constructing a `Bitset` of size `0` is unchanged.

**Comment by romintomasetti** [comment]
Hi @Rombur !

There is still a problem in the pipeline related to the helper structure in the test that makes the insert on device:
```bash
[ 95%] Building CXX object algorithms/unit_tests/CMakeFiles/Kokkos_AlgorithmsUnitTest_StdSet_Team_A.dir/TestStdAlgorithmsTeamForEachN.cpp.o

/var/jenkins/workspace/Kokkos/containers/unit_tests/TestUnorderedMap.hpp:450:9: error: Lambda passed to parallel_for implicitly captures this. [kokkos-implicit-this-capture,-warnings-as-errors]

        KOKKOS_CLASS_LAMBDA(const int) { this->m_map.insert(arg); });

        ^

/var/jenkins/workspace/Kokkos/core/src/setup/Kokkos_Setup_Cuda.hpp:57:29: note: expanded from macro 'KOKKOS_CLASS_LAMBDA'

#define KOKKOS_CLASS_LAMBDA [ =, *this ] __host__ __device__

                            ^

/var/jenkins/workspace/Kokkos/containers/unit_tests/TestUnorderedMap.hpp:450:9: note: the captured variable is used here.

        KOKKOS_CLASS_LAMBDA(const int) { this->m_map.insert(arg); });

        ^

/var/jenkins/workspace/Kokkos/core/src/setup/Kokkos_Setup_Cuda.hpp:57:34: note: expanded from macro 'KOKKOS_CLASS_LAMBDA'

#define KOKKOS_CLASS_LAMBDA [ =, *this ] __host__ __device__

                                 ^
```

I've added `NOLINTNEXTLINE(kokkos-implicit-this-capture)` to no avail. What should I do? Thanks!

**Comment by Rombur** [comment]
It looks like an issue with clang-tidy. Instead of using `this->m_map` try using `m_map` directly or create a local pointer to `m_map` in the function.

**Comment by romintomasetti** [comment]
Hi @Rombur !

Many tests failed for Cuda because of a driver mismatch. Can you ask for a re-test please?

Also, the CI on Windows is broken due to the helper `struct` for inserting values on device. Any idea of a fix? Is it a know issue?

**Comment by Rombur** [comment]
Retest this please

**Comment by romintomasetti** [comment]
Hi @Rombur !

> Also, the CI on Windows is broken due to the helper `struct` for inserting values on device. Any idea of a fix? Is it a know issue?

Still failing. Any thought?

**Comment by Rombur** [comment]
It looks like the compiler does not like `this->`. Are you sure you needed? You should be able to simply call `m_map`.

**Comment by Rombur** [comment]
clang-format is not happy

**Comment by romintomasetti** [comment]
@Rombur Now that all checks passed, it seems we need a 2nd approval to merge :smile:

**Comment by romintomasetti** [comment]
Hi @dalg24 ! Would you have a second to approve this one? @Rombur is already OK with it and the pipeline is green :wink: Thanks!

**Comment by romintomasetti** [comment]
> Hi @dalg24 @masterleinad ! Would you have a second to approve this one? @Rombur is already OK with it and the pipeline is green 😉 Thanks!

:up:

File: containers/src/Kokkos_Bitset.hpp [hunk 7]
```diff
@@ -70,13 +87,35 @@ class Bitset {
     block_shift = Kokkos::Impl::integral_power_of_two(block_size)
   };
 
+  //! Type of @ref m_blocks.
+  using block_view_type = View<unsigned*, Device, MemoryTraits<RandomAccess>>;
+
  public:
+  KOKKOS_DEFAULTED_FUNCTION
+  Bitset() = default;
```
[the remark this answers is not in the document]
Comment (romintomasetti) [reply containers/src/Kokkos_Bitset.hpp:95]: 1. Indeed, I can't remember why I made it default-constructible on the device. I guess I looked at how the default constructor was declared for `Kokkos::View` (which is also declared `KOKKOS_DEFAULTED_FUNCTION`). I don't have any strong opinion on that, so if you deem it useless that it be default-constructible on device, I can remove `KOKKOS_DEFAULTED_FUNCTION`.
2. I think I wanted to have - again - something similar to the `Kokkos::View`. If you don't give it a `arg_size`, default construct everything. Otherwise, initialize (see https://github.com/kokkos/kokkos/pull/6198#discussion_r1310465971 maybe). My point was really to make sure no allocation happens before the user really wants it to happen.

[the remark this answers is not in the document]
Comment (romintomasetti) [reply containers/src/Kokkos_Bitset.hpp:95]: 2. I guess it also does not make much sense to initialize the `Kokkos::Bitset` why a zero size?

[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_Bitset.hpp:95]: Ultimately, this is a drive-by change that changes behavior. Thus, I would prefer to address it in a different pull request.

[the remark this answers is not in the document]
Comment (romintomasetti) [reply containers/src/Kokkos_Bitset.hpp:95]: Are you asking just to revert `KOKKOS_DEFAULTED_FUNCTION` for the constructor or more than that? :smile:

[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_Bitset.hpp:95]: I would just leave this constructor's signature alone in this pull request.

[the remark this answers is not in the document]
Comment (romintomasetti) [reply containers/src/Kokkos_Bitset.hpp:95]: Ok, just to make my mind clear :smile:

So, instead of the current change:
```c++
  KOKKOS_DEFAULTED_FUNCTION
  Bitset() = default;

  /// constructor
  /// arg_size := number of bit in set
  explicit Bitset(unsigned arg_size) : Bitset(Kokkos::view_alloc(), arg_size) {}

  template <class... P>
  Bitset(const Impl::ViewCtorProp<P...>& arg_prop, unsigned arg_size)
```
I should do
```c++
  /// constructor
  /// arg_size := number of bit in set
  explicit Bitset(unsigned arg_size = 0u) : Bitset(Kokkos::view_alloc(), arg_size) {}

  template <class... P>
  Bitset(const Impl::ViewCtorProp<P...>& arg_prop, unsigned arg_size)
```
to stick with the previous behavior (`Bitset` always allocated, even with a zero size) and create an issue to discuss whether or not `Kokkos::Bitset` should be default constructed?

Sorry about the confusion :confused:

[the remark this answers is not in the document]
Comment (romintomasetti) [reply containers/src/Kokkos_Bitset.hpp:95]: Hell @masterleinad  ! Please don't forget me :angel:

[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_Bitset.hpp:95]: > I should do [...] to stick with the previous behavior (Bitset always allocated, even with a zero size) and create an issue to discuss whether or not Kokkos::Bitset should be default constructed?

Yes.

[the remark this answers is not in the document]
Comment (romintomasetti) [reply containers/src/Kokkos_Bitset.hpp:95]: * For the issue, there you go: #6467

**Comment by romintomasetti** [comment]
It seems the tests failed only for Cuda, because of a runner failure (`Disk quota exceeded`).

**Comment by romintomasetti** [comment]
> It seems the tests failed only for Cuda, because of a runner failure (`Disk quota exceeded`).

@masterleinad Would you mind launching the pipeline again? :smile:

**Comment by Rombur** [comment]
Retest this please

**Comment by romintomasetti** [comment]
Hi @masterleinad @dalg24 ! Would you have a minute for approving this one? Thanks!

File: core/src/OpenMP/Kokkos_OpenMP.cpp
Comment (masterleinad) [inline core/src/OpenMP/Kokkos_OpenMP.cpp:1]: Please split the drive-by changes moving `<iostream>` includes to a different pull request.

Comment (romintomasetti) [reply core/src/OpenMP/Kokkos_OpenMP.cpp:1]: Ok. It is #6482 :wink:

**Comment by romintomasetti** [comment]
I rebased on `Kokkos`'s develop branch because there was on issue for Cuda that is apparently resolved on the trunk.

**Comment by romintomasetti** [comment]
@masterleinad All checks passed :partying_face:

File: core/src/impl/Kokkos_ExecPolicy.cpp
Comment (masterleinad) [inline core/src/impl/Kokkos_ExecPolicy.cpp:1]: The changes here look unrelated.

Comment (romintomasetti) [reply core/src/impl/Kokkos_ExecPolicy.cpp:1]: Ouch indeed ! Let me repair this :wink:

Comment (romintomasetti) [reply core/src/impl/Kokkos_ExecPolicy.cpp:1]: Voilà :ballot_box_with_check:

File: core/src/impl/Kokkos_ExecPolicy.cpp (hunk 2 above)
Comment (romintomasetti) [reply core/src/impl/Kokkos_ExecPolicy.cpp]: Most of the includes related changes have been moved to #6482.

File: core/src/Threads/Kokkos_ThreadsExec.hpp (hunk 3 above)
Comment (romintomasetti) [reply core/src/Threads/Kokkos_ThreadsExec.hpp]: See #6482.

File: containers/unit_tests/TestBitset.hpp (hunk 6 above)
Comment (romintomasetti) [reply containers/unit_tests/TestBitset.hpp]: The changes related to `Bitset` allocation have been reverted, and a discussion opened in #6467.

File: containers/src/Kokkos_Bitset.hpp [hunk 8]
```diff
@@ -28,6 +28,23 @@
 
 namespace Kokkos {
 
+namespace Impl {
+//! Either append to the label if the property already exists, or set it.
+template <typename... P>
+auto with_updated_label(const ViewCtorProp<P...>& view_ctor_prop,
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/src/Kokkos_Bitset.hpp:34]: If we want to allow the use of custom labels with a fallback, I would do it this way.

[the remark this answers is not in the document]
Comment (romintomasetti) [reply containers/src/Kokkos_Bitset.hpp:34]: :+1:

**Comment by romintomasetti** [comment]
@dalg24 @masterleinad @Rombur The HPX action failed because of a timeout issue. Are we ready to merge this one anyway? :vertical_traffic_light:

**Comment by crtrott** [comment]
Thanks for your work on this! We appreciate it.
